selfservice/strategy/code/code_recovery_test.go

```diff
@@ -0,0 +1,91 @@
+package code_test
+
+import (
+	"database/sql"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/ory/kratos/internal"
+	"github.com/ory/kratos/selfservice/strategy/code"
+	"github.com/ory/kratos/x"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ory/x/stringslice"
+	"github.com/ory/x/urlx"
+
+	"github.com/ory/kratos/selfservice/flow"
+	"github.com/ory/kratos/selfservice/flow/recovery"
+)
+
+func TestRecoveryCode(t *testing.T) {
+	conf, _ := internal.NewFastRegistryWithMocks(t)
+
+	newCode := func(expiresIn time.Duration, f *recovery.Flow) *code.RecoveryCode {
+		return &code.RecoveryCode{
+			ID:        x.NewUUID(),
+			FlowID:    f.ID,
+			ExpiresAt: time.Now().Add(expiresIn),
+		}
+	}
+
+	req := &http.Request{URL: urlx.ParseOrPanic("https://www.ory.sh/")}
+	t.Run("func=GenerateRecoveryCode", func(t *testing.T) {
+		t.Run("case=creates unique tokens", func(t *testing.T) {
+			codes := make([]string, 100)
+			for k := range codes {
+				codes[k] = code.GenerateRecoveryCode()
+			}
+
+			assert.Len(t, stringslice.Unique(codes), len(codes))
+		})
+	})
+
+	t.Run("method=IsExpired", func(t *testing.T) {
+		t.Run("case=returns true if flow is expired", func(t *testing.T) {
+			f, err := recovery.NewFlow(conf, -time.Hour, "", req, nil, flow.TypeBrowser)
+			require.NoError(t, err)
+
+			token := newCode(-time.Hour, f)
+			require.True(t, token.IsExpired())
+		})
+		t.Run("case=returns false if flow is not expired", func(t *testing.T) {
+			f, err := recovery.NewFlow(conf, time.Hour, "", req, nil, flow.TypeBrowser)
+			require.NoError(t, err)
+
+			token := newCode(time.Hour, f)
+			require.False(t, token.IsExpired())
+		})
```

Comment:
What about a combination of flow is not expired, but code is? Or the other way around? Could that actually happen?

Reply:
This test only tests the method on the `RecoveryCode` struct, which does not take the flow into account.

The flow can expire before the code, in which case the flow will be invalidated first and the code is never checked.

In case the code expires first, this method is called. We have tests for both cases. 

- flow expired: https://github.com/ory/kratos/pull/2645/files#diff-d3976c6ed314fd1f2372a82a50ef612c974597c85bb49031eecd3f50952c3c40R829
- code expired: https://github.com/ory/kratos/pull/2645/files#diff-d3976c6ed314fd1f2372a82a50ef612c974597c85bb49031eecd3f50952c3c40R161

Also E2e: https://github.com/ory/kratos/pull/2645/files#diff-4156044df819e73198f46caad2012dc54e4d204a92f2417ff94d29f4a4676d61R75